pkg/kubelet/pluginmanager/pluginwatcher/README.md

```diff
@@ -1,9 +1,13 @@
 # Plugin Registration Service
 
 This folder contains a utility, pluginwatcher, for Kubelet to register
-different types of node-level plugins such as device plugins or CSI plugins.
-It discovers plugins by monitoring inotify events under the directory returned by
-kubelet.getPluginsDir(). We will refer to this directory as PluginsDir.
+different types of node-level plugins such as Device plugins, DRA plugins or
+CSI plugins. It discovers plugins by monitoring inotify events under the
+directory returned by kubelet.getPluginsDir(). We will refer to this directory
+as PluginsDir.
+
+Additionally, it periodically sends Registration.GetInfo gRPC requests to the
+Unix sockets found in this directory to monitor plugin health.
```

Comment:
A node may end up in a situation where the CSI driver pod looks working, but kubelet does thinks it's unhealthy. How can user figure that?

* Should we provide clear guidelines for the device / DRA / CSI plugins to monitor their end of the socket and do something when it gets closed?
* Is there a way how to recover from this situation? For example, restarting the CSI driver Pod could help, however, I don't see how kubelet could do it (it does not know what Pod created the socket).

Reply:
> Pods should be failing to run I guess when the driver is un-registered.

Isn't that impossible? The question here is about the situation where a plugin gets unregistered because it was failing to reply via gRPC. Not only is there no code in this PR which tries to tell the plugin that it got unregistered (or is there? - I haven't checked again), doing so at a time when the gRPC was unresponsive would probably not work.

Reply:
> Isn't that impossible? 

It's impossible to call an unregistered plugin if you're asking about it.

Reply:
So the question "Is there a way how to recover from this situation?" is still open because "Pods should be failing to run I guess when the driver is un-registered. " doesn't answer it, right?

Reply:
Nope. I can see one way to recover from this situation. This is what I described above - if pods are not running, cluster admin should be able to investigate it and restart the driver.

Reply:
The question was "A node may end up in a situation where the CSI driver pod looks working, but kubelet thinks it's unhealthy. ... Is there a way how to recover from this situation?".

In this case, the pod is running as far as the kubelet is concerned when monitoring the container, it's just not responding via gRPC long enough such that this new check unregisters the driver.

Will the node recover without manual intervention? If not, can the code be improved such that it does?

Reply:
> Will the node recover without manual intervention?

Yes, if driver starts to respond to GetInfo requests again, otherwise not.

> can the code be improved such that it does?

At least I don't see how to do it without changing registration API.

Reply:
Yes, there is a test `TestPluginStuckThenUnstuck` for this use case in this PR. I'm also working on e2e_node tests.

Reply:
Ack. I wrote https://github.com/kubernetes-csi/node-driver-registrar/issues/484 to check the CSI registrar that it behaves correctly when kubelet closes the socket.